### DISCUSSION TITLE
Computing correlation between two images using libvips

### DISCUSSION THREAD

**Discussion by nooblibvips** [libvips/libvips#2232] [Q&A]
All,

I'm trying to compute correlation between two very large pyramid tiffs. The resolution of bottom layer in each pyramid tiff is at least 40,000X80,000.I stumbled across [vips_fastcor](https://libvips.github.io/libvips/API/current/libvips-convolution.html#vips-fastcor) and below is my code to compute correlation between two images:

```
#include <stdio.h>
#include <vips/vips.h>
int main (int argc, char **argv)
{
  VipsImage *ref, *im, *corr_im;
  double avg_corr;

  if (VIPS_INIT (argv[0]))
    vips_error_exit ("unable to start VIPS");

  if (argc != 3)
    vips_error_exit ("usage: %s <reference image filename> <input file for computing correlation>", g_get_prgname ());

   if (!(ref = vips_image_new_from_file (argv[1], NULL)))
    vips_error_exit ("unable to open reference image");
  if (!(im = vips_image_new_from_file (argv[2], NULL)))
    vips_error_exit ("unable to open input file for computing QC");
 
  vips_fastcor(im, ref, &qc_im);
  if (vips_avg (qc_im, &avg_qc, NULL))
    vips_error_exit ("unable to find avg of QC image");
  
  printf ("Pixel average of QC image is %g\n", avg_qc);
  return (0);
```

My code is crashing when computing fast correlation. I have checked and libvips can read my input images. So, it is crashing only at the correlation step. Any pointers? 

Please let me know if you need additional information from me. Thanks in advance

**Comment by jcupitt** [comment]
Hello, `fastcor` searches for a small image within a big one, and for speed it maps the small image into memory before starting the search. Your program is probably running out of memory.

What do you need to compute? Is this is search problem?

**Reply by nooblibvips** [reply]
Thanks @jcupitt for clarifying what `fastcor` does.

Here's what am trying to do: In my experiments, I acquire a ground-truth image at time t=0. After the ground-truth image is acquired I process my object of interest and re-image it. I rinse and repeat this processing and acquisition loop at least 10 times. My hypothesis is that the processing does not affect the object of interest, and therefore image acquired at time t=10 will be the same as image acquired at time t=0. To prove my hypothesis, I'm computing the pixel-level correlation coefficient. My current benchmark is the following - 90% of pixels must have a correlation coefficient greater than 90.

Notes about input images:
I save t=0 image, and t>0 image as 16-bit monochrome tiled pyramid tiffs. I have to use the bottom layer of pyramid for computing correlation coefficient. Each tile is about 512x512 and the bottom layer is approximately 40,000x80,000 pixels.
Notes about what 'am doing now:
I'm currently using pyvips to read my images, extracting the bottom layer and computing correlation coefficient using python. I'm not running into memory issues with my python script. I'm working on converting my pyvips code to C/C++ when I stumbled across `fastcor`. Like you said, I cannot use it.

Are there functions in libvips that I can use? If not, do you have any other pointers?

Thanks

**Comment by jcupitt** [comment]
Yes, it doesn't sound like spatial correlation is right for your problem. Have you looked at things like spatial cielab?

https://web.stanford.edu/group/vista/cgi-bin/wiki/index.php/SCIELAB

Do you have some sample images? Not the whole slide, of course, just two cropped areas that you want to calculate a match for.

**Reply by jcupitt** [reply]
Or thinking about it, if you really do just want a single number for the match between two images, you can calculate the sum of squares of differences with something like:

```python
ssd = ((a - b) ** 2).avg() * a.width * a.height
```

which should run reasonably quickly.

**Comment by nooblibvips** [comment]
Thanks @jcupitt .

I have uploaded two sample images [here](https://danaherlifesciences.box.com/s/bz12b1io1vdbmgaa6tpzrduu6haxohms). This folder contains two images: (i) Round0 - my ground truth image and (ii) Round20 - image acquired after 20 rounds of processing to the slide. Like you asked, this is just a small field of view from my entire image. 

1. Why won't spatial correlation work for me? I'm trying to find which pixels have shifted between imaging rounds and which have been lost (either due to processing or focus issues in camera).
2. Thank you for suggesting SSD. It is a great metric for the entire image but I also need a metric at every pixel. My ultimate goal is to find which a pixel wise metric and eliminate those pixels in downstream analyses.

Thanks again for your immediate response. Please let me know if you need additional information or if I wasn't clear.

**Comment by jcupitt** [comment]
I suppose I don't understand what you mean by processing. Are you adding stains, perhaps, and then scanning the slide again? How are you resampling the scans to match?

You can just subtract and blur your sample images, then pick a difference threshold, is this the kind of thing you need?

![x](https://user-images.githubusercontent.com/580843/116577454-62f52e80-a908-11eb-9352-efd59850d830.jpg)

**Reply by nooblibvips** [reply]
Thanks @jcupitt. I'm sorry I uploaded the wrong images. I uploaded the correct images to the [same link](https://danaherlifesciences.app.box.com/folder/136500001659?s=bz12b1io1vdbmgaa6tpzrduu6haxohms) now.

Here's the processing am doing: (you are right) each time I bleach,  add 3-4 different stains, and image my slide using a multiplexed microscope. One of these 3-4 different stains is always a DAPI stain. The DAPI image acquired at t=0 serves as the ground-truth image, and all the other DAPI images acquired at t>0 are registered to the t=0 DAPI image. In theory, since 'am re-imaging the same slide repeatedly, I should not lose any cells. I'm therefore computing a spatial correlation coefficient to find the correlation between t=0 and t>0 DAPI images. I need to do this at a pixel level because I want to ignore any lost cells during segmentation, and phenotyping. 

Sorry again for uploading the wrong channel images earlier. Let me know if you need additional information.
